File: src/05.linkList/05.61.go
package linklist

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func RotateRight(head *ListNode, k int) *ListNode {
	// 暴力解法
	// n := 0
	// curNode := head
	// for curNode != nil {
	// 	curNode = curNode.Next
	// 	n++
	// }

	// if n == 0 {
	// 	return head
	// }

	// k %= n

	// if k == 0 || n == 1 {
	// 	return head
	// }

	// curNode = head
	// for i := 1; i < n-k; i++ {
	// 	curNode = curNode.Next
	// }

	// lastNode := curNode.Next
	// curNode.Next = nil

	// curNode = lastNode
	// for curNode.Next != nil {
	// 	curNode = curNode.Next
	// }
	// curNode.Next = head

	// return lastNode

	// 优化
	// var lastNode *ListNode
	// n := 0
	// curNode := head
	// for curNode != nil {
	// 	if curNode.Next != nil {
	// 		lastNode = curNode.Next
	// 	}
	// 	curNode = curNode.Next
	// 	n++
	// }

	// if n == 0 {
	// 	return head
	// }

	// k %= n

	// if k == 0 || n == 1 {
	// 	return head
	// }

	// curNode = head
	// for i := 1; i < n-k; i++ {
	// 	curNode = curNode.Next
	// }

	// newHead := curNode.Next
	// curNode.Next = nil
	// lastNode.Next = head

	// return newHead

	// 组成环
	if head == nil || head.Next == nil {
		return head
	}
	n := 1
	curNode := head
	for curNode.Next != nil {
		curNode = curNode.Next
		n++
	}

	m := n - k%n
	curNode.Next = head
	for m > 0 {
		curNode = curNode.Next
		m--
	}

	newHead := curNode.Next
	curNode.Next = nil

	return newHead
}

File: src/05.linkList/05.61.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

import { ListNode } from "./01.203";

export function rotateRight(head: ListNode | null, k: number): ListNode | null {
    // 暴力解法
    // let len = 0;
    // let curNode = head;
    // let lastNode = null
    // while (curNode) {
    //     if (curNode.next) {
    //         lastNode = curNode.next
    //     }
    //     curNode = curNode.next;
    //     len++;
    // }
    // k %= len;
    // if (!head || len === 1 || k === 0) {
    //     return head;
    // }

    // curNode = head;
    // for (let i = 1; i < len - k; i++) {
    //     curNode = curNode.next;
    // }

    // let newHead = curNode.next;
    // curNode.next = null;
    // lastNode.next = head;

    // return newHead;

    // 优化, 成为环
    if (!head || !head.next) {
        return head;
    }

    let len = 1;
    let curNode = head;
    while (curNode.next) {
        curNode = curNode.next;
        len++;
    }
    // 组成环
    curNode.next = head;

    let m = len - (k % len);
    while (m > 0) {
        curNode = curNode.next;
        m--;
    }
    let newHead = curNode.next;
    curNode.next = null;

    return newHead;
}
